db/src/lib.rs

```diff
@@ -0,0 +1,96 @@
+pub use sqlx;
+use sqlx::postgres::PgPoolOptions;
+pub use sqlx::PgPool;
+
+#[tracing::instrument(err)]
+pub async fn setup_db_pool() -> color_eyre::Result<PgPool> {
+    let database_url = std::env::var("DATABASE_URL").expect("DATABASE_URL must be set");
+    let pool = PgPoolOptions::new()
+        .max_connections(5)
+        .connect(&database_url)
+        .await?;
+    let mut connection = pool.acquire().await?;
+
+    let lock = sqlx::postgres::PgAdvisoryLock::new("knowledge-db-migration-lock");
+    let mut lock = lock.acquire(&mut connection).await?;
+
+    sqlx::migrate!().run(lock.as_mut()).await?;
+
+    lock.release_now().await?;
+    tracing::info!("Migration lock unlocked");
+
+    Ok(pool)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    mod test_db {
+        use sqlx::{migrate::MigrateDatabase, Postgres, Row};
+        use url::Url;
+
+        use super::*;
+        struct TestPool {
+            database_url: String,
+            pool: PgPool,
+        }
+
+        impl AsRef<PgPool> for TestPool {
+            fn as_ref(&self) -> &PgPool {
+                &self.pool
+            }
+        }
+
+        impl TestPool {
+            async fn teardown(self) -> color_eyre::Result<()> {
+                self.pool.close().await;
+                Postgres::drop_database(&self.database_url).await?;
+                Ok(())
+            }
+        }
+
+        async fn create_test_db_pool() -> color_eyre::Result<TestPool> {
```

Comment:
This is a helper you can use in tests! Instead of the 'real' `setup_db_pool` in specs we want to use this helper instead!

The difference is that this one will setup a temporary test DB to connect to, and delete that DB when the tests are done. That way different tests don't have a way to interact and leave the DB in a 'dirty' state!

So in a test that needs the DB you can call `create_test_db_pool` and then just remember to call `teardown` at the end to cleanup the test db!
[See example down below!]